### ISSUE TITLE
Make Scope.ParentScope public

### ISSUE THREAD

**Issue by dotnetjunkie** [simpleinjector/SimpleInjector#502]
`ParentScope` is currently an internal property of `Scope`, which enables `AsyncScopedLifestyle` and `ThreadScopedLifestyle` to do nesting, by iterating up the scope stack.

Making `ParentScope` public however, does considerably simplify creating a custom lifestyle that caches instances in the outer scope, for instance as described in #501.

**Comment by dotnetjunkie** [comment]
Won't be adding this feature. Even after making `ParentScope` public, there are many quirks to consider. Users are better of changing their design instead of depending on `ParentScope` to build a custom lifestyle.

**Comment by tompazourek** [comment]
I ran into a similar issue as #501 recently.

I was using the normal `AsyncScopedLifestyle` for the whole web request, but then I discovered a need to have some objects scoped to the current transaction. 

I'd still like to have some objects scoped to the outer `AsyncScopedLifestyle` scope (that corresponds to the current web request), and some objects I'd like to have scoped to the current transaction scope (part of the web request). Even inside the transaction scope, I still want to access some objects that are scoped to the outer (request) scope. As far as I know, it is not possible to do this just by using `AsyncScopedLifestyle`, as when I create a new scope (for the transaction), I cannot get instances scoped to the outer scope (the web request).

I solved this problem by creating two new scoped lifestyles: `OuterScopedLifestyle` and `TransactionScopedLifestyle` which both inherit from `AsyncScopedLifestyle`. However, I ran into two issues:

- The `ParentScope` is not public, so I need to get it through reflection.
- The Lifestyle `name` is not possible to set as there's no `AsyncScopedLifestyle` constructor that would accept a `name` (a protected one would be useful). I found the name useful for nicer error messages, if I don't change the name, both of my lifestyles are called "Async Scoped".

So far, I sorted it via reflection and it works as I need it to. But I think it would be beneficial to see the functionality around scoped lifestyles more open for extension. Right now, a lot of it is internal, and it's hard to build something custom that would work in async/await scenarios correctly.

---

@dotnetjunkie Any ideas on how to "change the design" of this? I need to have a request scope and sometimes a smaller transaction scope inside of that (but still access objects from the outer request scope). 

Or maybe reconsider to open this issue? Or provide more details about the quirks to consider that you mention?

Thanks!

**Comment by dotnetjunkie** [comment]
hi @tompazourek,

>  provide more details about the quirks to consider that you mention?

I think the gist of this is in [this 501 comment](https://github.com/simpleinjector/SimpleInjector/issues/501#issuecomment-356955233):

> _Refrain from using OuterScopedLifestyle, since it is unsuited for situations where the new operation runs in parallel, instead of simply asynchronously._

But do read the complete comment.

> Any ideas on how to "change the design" of this?

It's hard to comment on this in general terms, but it you show me what you got, I might be able to provide alternatives. What data/objects do you wish to pass on from scope to scope?

**Comment by tompazourek** [comment]
I tried to come up with an example that would illustrate my idea, and I fabricated this example.

Imagine this web request (inside a controller):

```csharp
ActionResult MarkOrderAsCompleted(int orderId)
{
	if (!this.CurrentUserManager.IsInRole(Roles.Administrator))
		throw new UnauthorizedAccessException();

	HandleCommand(new MarkOrderAsCompletedCommand { OrderId = 1 });	

	return RedirectToAction("...");
}
```

And imagine this handling (inside a command handler):

```csharp
void HandleCommand(MarkOrderAsCompleted command)
{
	this.MainDbConnection.DapperExecute(@"UPDATE [Orders] SET [State] = @completed, [ModifiedTimestamp] = @now, [ModifiedByUser] = @userId",
	new { completed = OrderStates.Completed, now = DateTimeOffset.Now, userId = this.CurrentUserManager.CurrentUserId });
}
```

A decorator wrapped this in a transaction, so the command handler is created inside a new scope.

In this scenario:

- The `CurrentUserManager` is scoped to the request. In my example, it checks the permissions which could possibly be moved inside the command handling, but it can be used for other things (for example, we could use it to show different things in the UI view depending on current user's role). It's also used inside the transaction, but I'd prefer to share the same instance as exists outside of the transaction (why have two instances when we need only one?).
- The `MainDbConnection` is scoped to the transaction. That way, we get a new connection (with a transaction linked to it) every time we try to execute new command from the request. If we'd execute two transactions in the request (e.g. if the first transaction one fails, we want to go save the information somewhere in a second transaction), each of these would get new instance of `MainDbConnection`.

If I would only use `AsyncScopedLifestyle` and created new scope in the decorator around my command handler, I would get a new instance of `CurrentUserManager` inside my command handler. However, that's not necessary, I am perfectly fine sharing the instance that exists outside of the transaction.

In fact, the majority of my objects will be request-scoped. It's only the objects that deal with the current transaction that will be transaction-scoped. These are usually objects that use a database connection that supports transactional behavior. So it doesn't make sense to me to recreate all the objects in the inner transaction scope when I can re-use most of them from the request scope.

Sorry if I'm not explaining this clearly, I very much appreciate your help.

---

To implement this, I created two new lifestyles:

- `OuterScopedLifestyle`
	- When an object has this lifestyle, I ignore the scopes that are marked as transaction scopes, and find the closest parent scope that's not a transaction scope.
- `DomainTransactionScoped`
	- When an object has this lifestyle, I find the top-most transaction scope. So if there are multiple transaction scopes nested, I only consider the top one. I also have an option here with fallback. Some objects are scoped to the transaction, but if there's no transaction, they can still be scoped to the outer scope.

This is the code I have:

```csharp
public class OuterScopedLifestyle : AsyncScopedLifestyle
{
    #region Main

    private static readonly object OuterScopeIdentifierKey = "__" + nameof(IsOuterScope);
    private static readonly object OuterScopeIdentifierValue = true;

    public OuterScopedLifestyle()
    {
        this.SetName("Outer Scoped");
    }

    protected override Func<Scope> CreateCurrentScopeProvider(Container container)
    {
        var currentScopeProvider = base.CreateCurrentScopeProvider(container);
        return () =>
        {
            var currentScope = currentScopeProvider();
            var closestParentNonDomainTransactionScope = GetClosestParentNonDomainTransactionScope(currentScope);
            return closestParentNonDomainTransactionScope;
        };
    }

    protected override Scope GetCurrentScopeCore(Container container)
    {
        var currentScope = base.GetCurrentScopeCore(container);
        var closestParentNonDomainTransactionScope = GetClosestParentNonDomainTransactionScope(currentScope);
        return closestParentNonDomainTransactionScope;
    }

    /// <summary>
    /// Returns current scope regardless of if it's domain transaction scope or outer scope.
    /// </summary>
    public Scope GetCurrentScopeAny(Container container)
    {
        var currentScope = base.GetCurrentScopeCore(container);
        return currentScope;
    }

    public new static Scope BeginScope(Container container)
    {
        var scope = AsyncScopedLifestyle.BeginScope(container);
        scope.SetItem(OuterScopeIdentifierKey, OuterScopeIdentifierValue);
        return scope;
    }

    private Scope GetClosestParentNonDomainTransactionScope(Scope scope)
    {
        if (scope == null)
            return null;

        do
        {
            if (!DomainTransactionScopedLifestyle.IsDomainTransactionScope(scope))
                return scope;

            scope = scope.GetParentScope();
        } while (scope != null);

        return null;
    }

    internal static bool IsOuterScope(Scope scope)
        => scope.GetItem(OuterScopeIdentifierKey) == OuterScopeIdentifierValue;

    #endregion
}
```

```
public class DomainTransactionScopedLifestyle : AsyncScopedLifestyle
{
    #region Main

    private static readonly object DomainTransactionScopeIdentifierKey = "__" + nameof(IsDomainTransactionScope);
    private static readonly object DomainTransactionScopeIdentifierValue = true;

    /// <remarks>
    /// Prevents outer-scoped objects depending on domain-transaction-scoped.
    /// Also prevents domain-transaction-with-fallback-scoped objects depending on domain-transaction-without-fallback-scoped
    /// objects.
    /// </remarks>
    public override int Length => FallbackToOuterScope
        ? Scoped.Length - 1
        : Scoped.Length - 2;

    public DomainTransactionScopedLifestyle(bool fallbackToOuterScope)
    {
        FallbackToOuterScope = fallbackToOuterScope;
        this.SetName(FallbackToOuterScope ? "Domain Transaction Scoped With Fallback" : "Domain Transaction Scoped");
    }

    public bool FallbackToOuterScope { get; }

    protected override Func<Scope> CreateCurrentScopeProvider(Container container)
    {
        var currentScopeProvider = base.CreateCurrentScopeProvider(container);
        return () =>
        {
            var currentScope = currentScopeProvider();
            var topMostParentDomainTransactionScope = GetTopMostParentDomainTransactionScope(currentScope);

            if (FallbackToOuterScope && topMostParentDomainTransactionScope == null)
                return currentScope;

            return topMostParentDomainTransactionScope;
        };
    }

    protected override Scope GetCurrentScopeCore(Container container)
    {
        var currentScope = base.GetCurrentScopeCore(container);
        var topMostParentDomainTransactionScope = GetTopMostParentDomainTransactionScope(currentScope);

        if (FallbackToOuterScope && topMostParentDomainTransactionScope == null)
            return currentScope;

        return topMostParentDomainTransactionScope;
    }

    /// <summary>
    /// Returns current scope regardless of if it's domain transaction scope or outer scope.
    /// </summary>
    public Scope GetCurrentScopeAny(Container container)
    {
        var currentScope = base.GetCurrentScopeCore(container);
        return currentScope;
    }

    public new static Scope BeginScope(Container container)
    {
        var scope = AsyncScopedLifestyle.BeginScope(container);
        scope.SetItem(DomainTransactionScopeIdentifierKey, DomainTransactionScopeIdentifierValue);
        return scope;
    }

    private Scope GetTopMostParentDomainTransactionScope(Scope scope)
    {
        if (scope == null)
            return null;

        Scope topMostParentCommandScope = null;
        do
        {
            if (IsDomainTransactionScope(scope))
            {
                topMostParentCommandScope = scope;
            }

            scope = scope.GetParentScope();
        } while (scope != null);

        return topMostParentCommandScope;
    }

    internal static bool IsDomainTransactionScope(Scope scope)
        => scope?.GetItem(DomainTransactionScopeIdentifierKey) == DomainTransactionScopeIdentifierValue;

    #endregion
}
```

The outer scope is created using `OuterScopedLifestyle.BeginScope(container)`, the transaction scope is created inside of the outer scope using `DomainTransactionScopedLifestyle.BeginScope(Container)`.

So far, I haven't encountered any issues with my solution, it works exactly how I want it to work (that doesn't mean I don't encounter issues in the future, of course).

My only issue is that I need to use the reflection:

```
internal static class SimpleInjectorReflectionExtensions
{
    private static readonly PropertyInfo _parentScopeProperty = typeof(Scope)
        .GetProperty("ParentScope", BindingFlags.Instance | BindingFlags.NonPublic);

    private static readonly FieldInfo _lifestyleNameBackingField = typeof(Lifestyle)
        .GetField("<Name>k__BackingField", BindingFlags.Instance | BindingFlags.NonPublic);

    public static Scope GetParentScope(this Scope scope)
    {
        if (_parentScopeProperty == null)
            throw new InvalidOperationException("Cannot access the Simple Injector's Scope.ParentScope property via reflection.");

        var parentScope = (Scope) _parentScopeProperty.GetValue(scope);
        return parentScope;
    }

    public static void SetName(this Lifestyle lifestyle, string name)
    {
        if (_lifestyleNameBackingField == null)
            throw new InvalidOperationException("Cannot access the Simple Injector's Lifestyle.Name backing field via reflection.");

        _lifestyleNameBackingField.SetValue(lifestyle, name);
    }
}
```

**Comment by TheBigRic** [comment]
As @dotnetjunkie stated in one of the first comments. It seems that there is almost always a better solution by changing your design than to fall back to this outer/inner scope implementation. This complicates your code. 

And this is case for the example you supplied. With some simple design changes you'll easily fix the issues you ran into without complexing your code. 

`CurrentUserManager` can be made a singleton if you take the state out of it. Currently is not completely clear to me if your current implementation has state but based on your shown usage this implementation can be made singleton.

``` c#
public class HttpContextWebUserContext : IWebUserContext
{
    public IIdentity CurrentUser => HttpContext.Current.User.Identity;
    public bool IsInRole(Roles role)
    {
       // determine if user is role based on cookie or otherwise;
     return someDecision;
    }
}
```

The other point you bring to the table is handling multiple commands from a single ActionMethod. There are some pretty need solutions to this problem. @qujck wrote [this](http://scrapbook.qujck.com/holistic-abstractions-take-2/) interesting blogpost on how to solve this. 

In summary, use at 'ICommandHandler<>' as some kind umbrella under which you can handle multiple `ISpecificCommandHandler<>`. `ICommandHandler<>` can in that case be decorated with a:

- `CheckPermissionCommandHandlerDecorator<>`
- `TransactionScopeCommandHandlerDecorator<>`
- `SaveChangesCommandHandlerDecorator<>`

While you can decorate you `ISpecificCommandHandler<>` with specific 'ValidationCommandHandlerDecorators' or whatever. 

As @qujck states: 

> Commands and Queries are Holistic Abstractions. Holistic Abstractions are concerned with the whole transaction; they are not a part of a transaction

With this design there is absolutely no need at all to create inner and outer scopes.

**Comment by tompazourek** [comment]
@TheBigRic Thanks for your perspective on this and also sharing some interesting links. I always like to think about software architecture and get other people's experiences and views, especially when it's challenging my opinions.

You argue that:

> With this design there is absolutely no need at all to create inner and outer scopes.

But you're still mentioning `TransactionScopeCommandHandlerDecorator`. If a decorator creates a transaction scope, it is a new scope inside the outer request scope.

In my opinion, there are still objects whose lifetime depends on the transaction scope:

- A database transaction object depends on the transaction scope.
- A database connection object depends on the database transaction.
- A unit of work object (like EF's `DbContext`) depends on the database connection and current transaction.

These objects have their lifetime determined by the lifetime of the transaction scope. If I create a `DbContext`, and link it to a database transaction, when the transaction is finished, I can no longer use that `DbContext`.

Unless I explicitly say that each request is always a single transaction (and by doing so merging those two scopes), there will be some differences. And I don't think it's correct to assume that each request is exactly one transaction, I prefer if there is a difference between those.

What's your opinion on this? When I look at some of these objects, I really do see that their lifetime depending on the lifetime of the transaction, and not on the lifetime of the whole request. Is that not the case?

**Comment by TheBigRic** [comment]
>  And I don't think it's correct to assume that each request is exactly one transaction, I prefer if there is a difference between those.

I think I disagree on this point. For me each request **must** be a single transaction. If we would use 2 transactions within in a single request problems would arise fairly quick. Let's say the first transaction succeeds and thereby data is modified. Than a second transaction fails, what will you do to inform the user? What must you do to fix your broken database state? A definite no-go for me. 

This does not mean you can't handle multiple commands, as already shown.

For me, it also does not mean you could use a second TransactionScope in the same request to perform a query, but not a command! This will be wrapped (in my case) in a new AsyncScope and will create new `DbContext`, etc.

**Comment by tompazourek** [comment]
@TheBigRic I understand your point with multiple transactions. I agree that a single transaction per request is usually more than enough. I've encountered some use-cases where I'd want to execute a second transaction only if the first transaction fails. If the first transaction deals with external systems that are not fully transactionally-aware, and it fails, I might want to follow the failure with a second transaction to save the error and mark the relevant item that it needs a manual review, for example. I myself don't dismiss it completely, I think there are some rare cases when it makes sense.

But even when we accept the limit and insist on a single transaction per request, I still believe it makes sense to have control over when the transaction starts and when it ends. For example, sometimes I make queries that just read from the database, and they don't need to be in an explicit transaction. Some GET requests are all about reading, and they might not need to deal with the overhead of locking and isolation, so they might not need to be wrapped in an explicit transaction. However, if it's a query that's run inside a command, I prefer for it to be in a transaction, so that appropriate locking and isolation behaviour will occur. At the moment, I like to only start the explicit transaction around commands (and subsequently all queries that run inside the command handlers), I don't wrap every query or request in an explicit transaction.

You mentioned that in your case you create a new `AsyncScope`. My issue is that when I create this new `AsyncScope`, it creates all the objects again. Some objects are in my opinion perfectly fine to be scoped to the outer scope. I don't think I need to recreate everything just because I need a new `DbContext`. What if I have a request-scoped cache that holds some state as a performance optimization? I cannot re-create this in the inner scope without losing all that state.

Could you maybe elaborate a bit on when you wrap things in a new `AsyncScope` and why? I'm not sure if I follow completely, and I'm very interested in your perspective.

**Comment by dotnetjunkie** [comment]
Hi @tompazourek,

I understand that you want to run your command handlers in an isolated bubble. I think there is nothing wrong with that. I think that in general, you want to have one transaction per operation, but that doesn't mean you should wrap the transaction around the whole request. I actually consider that to be a bad thing. Instead, you would always wrap the transaction around the command handler. I know this is something that Ric does as well :).

As you described, when the transaction fails, both the connection and DbContext are doomed. You don't want to reuse that same DbContext outside the transaction and companioning TransactionScope. This is one reason why you shouldn't wrap the transaction around the complete web request.

What I fail to see, however, is why you *need* to reuse some registrations across scopes within the same request. You say "why have two instances when we need only one?," but I fail to see why having one instance per scope would be problematic? I would like to turn the question around: "why not?" Can you elaborate on what your `CurrentUserManager` does internally that would force you to reuse it? Can you show this code?

Do note that even if we would make the `Scope.ParentScope` property public, I think your solution that depends on `Scope.ParentScope` will always be more complicated compared to a solution with just one `CurrentUserManager` per scope. It does mean, however, that you need to share underlying runtime data in a different way. Ric showed one solution, but there are others, which is something we can talk about. But I'd like you to start with explaining what  `CurrentUserManager` does that makes you want to use this stacked scoping model.

**Comment by tompazourek** [comment]
@dotnetjunkie Thank you for responding, I appreciate your comments.

I have been thinking about this in the meantime and I think I came to a similar conclusion. I think all the confusion has to do with the dependencies of my command handlers.

The `CurrentUserManager` object that I gave as the example does load additional information relevant to the currently logged-in user (like their user settings and other data). It is actually a form of caching, I load the data the first time I need it, and then it's just stored until the end of the request. I sometimes use it in a query, sometimes in a command, sometimes in a view...

If I'd use a brand new `AsyncScope`, this data won't be cached and will be loaded again when I am inside the command handler. I think that is just wasteful - why load the data twice when I don't need to? A way to solve this is to start passing more data in the command itself (so I'd lose the dependency on `CurrentUserManager` in the command handler), but that doesn't feel right to me either. Some of this data is an implementation detail and shouldn't belong in the interface (i.e. the command).

I have other objects that work in a similar way, basically a request-scoped caches. I use it as a performance optimization in a variety of cases.

I also have some objects that communicate with REST APIs and use a HttpClient. I prefer to use the same HttpClient for the whole request as I believe it can keep the connection open if there are multiple API requests. I don't like to create new one just because I'm executing a command, there's just additional work involved in it.

Then I have some objects that are linked to the request itself. Like that I generate a unique identifier for the current request, and I store the timestamp of when the request started. It's basically state linked to the current request. I use this data in some command handlers as an implementation detail. But if command handlers run in their isolated bubble, I lose access to all of this.

I hope I'm explaining this clearly. I tried to give 3 separate examples, but maybe they are actually the same thing in principle.

Have you ever dealt with something like this?

**Comment by dotnetjunkie** [comment]
> If I'd use a brand new AsyncScope, this data won't be cached and will be loaded again when I am inside the command handler. I think that is just wasteful

I agree; this could be wasteful.

> why load the data twice when I don't need to?

Agreed. If this causes a performance problem, request-based caching is a good solution.

> A way to solve this is to start passing more data in the command itself (so I'd lose the dependency on CurrentUserManager in the command handler), but that doesn't feel right to me either. Some of this data is an implementation detail and shouldn't belong in the interface 

Again, you are completely right here. Those implementation details should not be passed along through the public API of the system; that would pollute the API and can easily cause maintainability issues.

But `Scope` instances are not the only way to implement caching. @TheBigRic already gave a great example of how you can use `HttpContext.Current` as a way to implement caching. To be more precise, the `HttpContext.Current.Items` property gives access to a request-specific dictionary that functions as a cache. You can very easily use as cache instead of storing the values in private fields of the class. When doing this, you will be able to still have all nested scopes run in an isolated bubble, while sharing some request-specific data that you explicitly control.

> Have you ever dealt with something like this?

Yes, exactly as I described above. Use `HttpContext.Current.Items`. Or in case you're running ASP.NET Core, use `IHttpContextAccessor` and call `IHttpContextAccessor.HttpContext.Items` with the same effect.

**Comment by tompazourek** [comment]
@dotnetjunkie You're right, it actually would be possible to use `HttpContext.Current.Items` to work around this.

I am not actually sure if I like this approach better, it doesn't feel quite right to me (but my intuition might be wrong). At least I now see there's a way to do this with just the included `AsyncScopedLifestyle`. I still like the idea of having a smaller scope just for the transaction-related objects. I'll have to think about if it's worth pursuing.

Are there any other advantages to having these fully isolated bubbles for command handlers (besides not having to invent custom lifestyles)?

**Comment by dotnetjunkie** [comment]
Hi Tom,

There are many levels on which you can decide how to cache values:

* Don't cache at all
* Cache values for the duration of a single method call
* Cache values within a single class
* Cache values for the duration of a single Simple Injector scope
* Cache values for the duration of a web request
* Cache values linked to a user's session.
* Cache values in-memory for the lifetime of the app domain (possibly with expiration)
* Cache values out-of-process to be reused over machines, for instance using SQL Server, MemCache, Redis, or another temporary store.
* etc

There are many options here and some options can be used interchangably without having any functional effect on the system. For instance, any options is interchangable with the 'Don't cache at all' option. Requests might take longer, but the functional outcome of the system is identical.

In your case, whether you choose to use caching on scoping level or on request level does not functionally change the system. In other words, changing from scope-level caching to request-level caching is just a performance optimization. You might as well use a session-based, out-of-process cache, with the same functional effect, and it might even be more performant.

So from that sense, the use of `HttpContext.Current.Items` as a form of request-level caching is not a "workaround" at all. It is simply a possible solution to implement caching. In the specific case where request-level caching is the most suitable level of caching, you happen to have two options to choose from: outer-scoping and `HttpContext.Current`. If you go a level up in caching, it will become impossible to use Simple Injector scopes for scoping, and you will have to use constructs such as `HttpContext.Current` or classes like Microsoft's `MemoryCache` to get a appdomain-wide cache.

So as I see it, you have two equaly valid choices here, where one options is very easy to implement (using `HttpContext.Current` or a similar construct depending on whether you are using ASP.NET or ASP.NET Core), while the other options (using outer-scopes) is rather painful to implement.

Pick your battles ;-)

**Comment by tompazourek** [comment]
Hi Steven,

In terms of caching, you're right about this. It is just one of the levels, and once you put in that perspective, it becomes clear that it is a good way to think about it. So thank you for this idea.

I still like the idea of having objects scoped to the lifetime of the request and some objects scoped to the lifetime of the transaction, even if I were to abandon this as a form of caching. But it might be possible to just do this via the `AsyncScope` wrapping decorator around command handlers.

---

One of the things that my `DomainTransactionScopedLifestyle` did was that it always found the top-most domain transaction scope. It might be a bit difficult to explain, so please bear with me. I wrap every command handler with a scope via a decorator (similarly as @TheBigRic suggested). But my command handlers sometimes execute other commands inside of them (some of my commands do one thing, some of my commands do more things, so I reuse the code by calling other commands inside). And all of these share the same domain transaction scope. I find this difficult to do with just the `AsyncScope`, there'll be potentially multiple scopes nested inside each other (one for each command that I nest inside another command), but I still need to share the same database connection and transaction (and `DbContext`).

I suppose one way one can deal with this is by preventing command handlers from executing other commands inside. Basically that each command equals a single transaction. But that feels very restrictive, especially if I already limit myself to a single command per request. If I want to do multiple different things all in a single transaction (e.g. update two completely different entities at once), I'll have to have one more layer behind command handlers just so that I can reuse code.

Or is there some other way? Any thoughts on this (or links to read)?

**Comment by dotnetjunkie** [comment]
> Basically that each command equals a single separate transaction and no nesting occurs. But that feels very restrictive

Sure. It is a restriction. But in a sense, all architectural decisions are restrictions to our software. Restriction that try to find the right balance in maintainability.

I've been practicing this model using commands for quite some years now, and by now I've helped a large number of developers with this architecture. Question about how to deal with cross-cutting concerns when commands are nested, is one question that comes up quite regularly. As you already noticed, nesting commands can complicate the application of cross-cutting concerns tremendously. I found out a long time ago that adding the nesting-restriction typically makes the system less complex, and easier to comprehend, even though command reuse goes out the window. I think it's a reasonable architectural decision.

Not nesting commands, however, has other interesting benefits. From a design perspective, for instance, it becomes much clearer what a command is. In that case a command becomes a *business transaction*—a use case. I find it much easier to explain to the business, and to communicate with other developers about it.

When the system becomes more complex, though, you want to start reusing application logic. Having an architectural rule that states that commands can't be nested, is no excuse for introducing all kinds of code duplication inside your commands.

In the most basic form, you can use the [Extract Method refactoring](https://refactoring.com/catalog/extractMethod.html) to move code to a reusable spot. Sometimes you can create a static class with a static helper method, but often times you will need to go beyond this simple code extraction technique and apply other patterns. Here are some things that come to mind.

* You can hide duplicated logic with their dependencies behind a new service abstraction. This pattern is known as the [Facade Services refactoring](https://blog.ploeh.dk/2010/02/02/RefactoringtoAggregateServices/). In [my book](https://manning.com/seemann2) I actually spend a fair amount of text on this refactoring (see section 6.1).
* When you have code that needs to be executed for some use case, but isn't really part of the core operation, you might want to introduce events and event handlers. This introduces a new `IEventHandler<TEvent>` abstraction, with an identical signature as your `ICommandHandler<TCommand>` abstraction, but with completely different semantics. You could have multiple handlers responding to the same `TEvent`, and a command handler can raise such event. You can decide to run these handlers within the same transaction, for instance after the command handler has run, but before the operation is committed. Or you can store these events in a durable queue. This allows you to run the operations at a later time. This allows the main operation (the command) to succeed, even if some external system is down, or prevent slow operations from having the user have to wait. This even allows you to introduce retry mechanism, which improves the availability of the system. There's a lot been written about this, but again, section 6.1 of my book touches this subject. Another great introduction to messaging and queueing is [Learning NServiceBus Second Edition by David Boike](https://www.packtpub.com/application-development/learning-nservicebus-second-edition). The book is very informative, even if you're not interested in using NServiceBus.
* Perhaps part of the duplication is validation of input or execution of business rules. You might want to introduce new abstraction for this. For instance an `IValidator<T>` abstraction allows you to validate command logic or other messages, and those validators can be kicked off by a command-handler decorator. Or you can define an `IBusinessRule<TEntity>` that can be triggered (by a command-handler decorator) when an entity is changed. You could even split these rules in entity-validation and rules that actually mutate the entity before it’s persisted.
* Events are really interesting for "execute-after" kind of scenarios. Somethimes, however, you want to split the main operation into multiple distinct steps. @TheBigRic already aluded at this, but you can introduce a new abstraction for this, a la `ISubCommandHandler<T>`. This is described by @qujck [here](http://scrapbook.qujck.com/holistic-abstractions-take-2/). To me, those `ISubCommandHandler<T>` abstraction get especially useful if you wish to apply cross-cutting concerns on that level as well.

In the end, it's all about defining [holistic abstractions](http://scrapbook.qujck.com/holistic-abstractions-take-2/), as @qujck put it. This helps you in having a clear definition what an abstraction and its messages are. So you can have:
* `ICommandHandler<TCommand>` as definition of a complete business transaction.
* `IQueryHandler<TQuery, TResult>` as [definition of a query](https://blogs.cuttingedge.it/steven/posts/2011/meanwhile-on-the-command-side-of-my-architecture/).
* `IValidator<T>` as definition of some piece of input validation.
* `IBusinessRule<TEntity>` as definition of some piece of entity-specific business rule.
* `IEventHandler<TEvent>` as definition for some logic that that needs to handle a specific business event, while needs to be running in isolation of other operations, and possibly need to be retried.

I hope this helps.

**Comment by tompazourek** [comment]
Thanks for the long response, it's very helpful. It makes me challenge some of my views about architecture. I appreciate other views on this, and also the links you posted with more details. I'll have to take a look at your book (I see it's already finished, so congratulations), and the NServiceBus book also does look very interesting. I haven't used a service bus before, I've only used Hangfire for some of its use-cases, but it certainly is an interesting topic to explore.

I've read the techniques you mention, and I think the issues where I think nesting commands makes sense is the first one (deduplicating logic behind a new service abstraction). I have been enjoying the simplicity of CQS as the design approach to the domain layer. You don't need to think about how many services you're going to have, what will the methods look like, which methods modify data, which methods don't. Everything is either a command or a query, it's very simple. Which is why I haven't designed the domain layer as services (and facade services), but as commands/queries and their handlers. I use services for the data access layer, but there it's not about domain logic, but more about connecting to the data sources and data mapping. I have used this CQS approach on a fairly large web application (~300 database tables), and I find it very maintainable, it's clear how to write new code, and how to modify the existing, so I'm definitely a big fan.

I suppose when you're using one more abstraction (the domain services) behind the command handlers, it means that your commands actually mean something different than my commands without the domain services behind. My commands are simply actions that somehow change the state of the domain, and the transaction aspect is beyond them and is handled outside. Where in the case you describe the commands *are* the transactions, which is also an interesting way to think about it, but quite different.

I'll give an example for when I've used nesting commands:

- I have a command that is written for a form where a user edits the prices of a product, e.g. `UpdateProductPricesCommand`.
- I have a command that is written for a form where a user edits the SKU codes of a product (MPNs, GTINs, barcodes), e.g. `UpdateProductStockCodesCommand`.
- We get a new feature where we import products from a third party system. And either all of the data for a single product gets saved, or none of it (so it's simple to keep track of how successful the import was). So I've got a new command, e.g. `ImportProductCommand`. 
- The command handler for `ImportProductCommand` executes both `UpdateProductPricesCommand` and `UpdateProductStockCodesCommand` inside of it (so it's in a single transaction).

Now if I don't want to implement this via nesting commands, I have to have one more layer in my application so that I can deduplicate the code. But then I end up with commands, queries, and domain services, and that's where I think I'm losing the simplicity, I'll need to think about which code belongs where, and put more thought into how to structure it, and with services it's not as clear which methods are read-only and which are changing the state. I don't like domain services because I think it's not as straightforward to write, and also it feels like too many layers of abstraction to me (but that's just my intuition) and too many types of different shapes and purposes.

I suppose the approach to this also depends on what kind of system is being implemented and what are its requirements.

---

Now back to the topic of this issue:

Would you be open to changes regarding:

1. Changing the `ParentScope` property from `internal` to `public`,
2. Adding a `protected` constructor to `AsyncScopedLifestyle` to be able to set the `name` in child classes. Without the `name` constructor, the class might as well be `sealed` as when you inherit from it, you'll most likely want to have a new name.

I could contribute with a PR to this (or maybe two separate PRs if you'd prefer), but would you allow these changes?

I think there might be other scenarios where someone might be willing to tweak with the scoped lifestyles, and I don't see a reason why to make the extension intentionally difficult. I know that one of the principles of Simple Injector is to promote good practices to steer developers towards the proven way of doing things, which I very much appreciate, but I don't think that making this change is actually in violation of that.
